src/topology/metric_space/infsep.lean

```diff
@@ -0,0 +1,464 @@
+  { rw not_nontrivial_iff at hs, exact hs.infsep_zero }
+end
+
+lemma nontrivial.infsep_of_fintype [decidable_eq α] [fintype s] (hs : s.nontrivial) :
+  s.infsep = s.off_diag.to_finset.inf' (by simpa) (uncurry dist) :=
+by { classical, rw [infsep_of_fintype, dif_pos hs] }
+
+lemma finite.infsep [decidable s.nontrivial] (hsf : s.finite) :
+  s.infsep = if hs : s.nontrivial then hsf.off_diag.to_finset.inf' (by simpa) (uncurry dist)
+  else 0 :=
+begin
+  split_ifs with hs,
+  { refine eq_of_forall_le_iff (λ _, _),
+    simp_rw [hs.le_infsep_iff, imp_forall_iff, finset.le_inf'_iff, finite.mem_to_finset,
+            mem_off_diag, prod.forall, uncurry_apply_pair, and_imp] },
+  { rw not_nontrivial_iff at hs, exact hs.infsep_zero }
+end
+
+lemma finite.infsep_of_nontrivial (hsf : s.finite) (hs : s.nontrivial) :
+  s.infsep = hsf.off_diag.to_finset.inf' (by simpa) (uncurry dist) :=
+  by { classical, simp_rw [hsf.infsep, dif_pos hs] }
+
+lemma _root_.finset.coe_infsep [decidable_eq α] {s : finset α}
+  : (s : set α).infsep = if hs : s.off_diag.nonempty then s.off_diag.inf' hs (uncurry dist)
+                         else 0 :=
+begin
+  have H : (s : set α).nontrivial ↔ s.off_diag.nonempty,
+  by rwa [← set.off_diag_nonempty, ← finset.coe_off_diag, finset.coe_nonempty],
+  split_ifs with hs,
+  { simp_rw [(H.mpr hs).infsep_of_fintype, ← finset.coe_off_diag, finset.to_finset_coe] },
+  { exact ((not_nontrivial_iff).mp (H.mp.mt hs)).infsep_zero }
+end
+
+lemma _root_.finset.coe_infsep_of_off_diag_nonempty [decidable_eq α] {s : finset α}
+  (hs : s.off_diag.nonempty) : (s : set α).infsep = s.off_diag.inf' hs (uncurry dist) :=
+by rw [finset.coe_infsep, dif_pos hs]
+
+lemma _root_.finset.coe_infsep_of_off_diag_empty [decidable_eq α] {s : finset α}
+  (hs : s.off_diag = ∅) : (s : set α).infsep = 0 :=
+by { rw ← finset.not_nonempty_iff_eq_empty at hs, rw [finset.coe_infsep, dif_neg hs] }
+
+lemma nontrivial.infsep_exists_of_finite [finite s] (hs : s.nontrivial) :
+∃ (x y ∈ s) (hxy : x ≠ y), s.infsep = dist x y :=
+begin
+  classical,
+  casesI nonempty_fintype s,
+  simp_rw hs.infsep_of_fintype,
+  rcases @finset.exists_mem_eq_inf' _ _ _ (s.off_diag.to_finset) (by simpa) (uncurry dist)
+    with ⟨_, hxy, hed⟩,
+  simp_rw mem_to_finset at hxy,
+  exact ⟨w.fst, hxy.1, w.snd, hxy.2.1, hxy.2.2, hed⟩
+end
+
+lemma finite.infsep_exists_of_nontrivial (hsf : s.finite) (hs : s.nontrivial) :
+∃ (x y ∈ s) (hxy : x ≠ y), s.infsep = dist x y :=
+by { letI := hsf.fintype, exact hs.infsep_exists_of_finite }
+
+end pseudo_metric_space
+
+section metric_space
+variables [metric_space α] {x y z: α} {s t : set α}
+
+lemma infsep_zero_iff_subsingleton_of_finite [finite s] :
+  s.infsep = 0 ↔ s.subsingleton :=
+begin
+  rw [infsep_zero, einfsep_eq_top_iff, or_iff_right_iff_imp],
+  exact λ H, (einfsep_pos_of_finite.ne' H).elim
+end
+
+lemma infsep_pos_iff_nontrivial_of_finite [finite s] :
+  0 < s.infsep ↔ s.nontrivial :=
```

Comment:
can you also give these lemmas for `(s : finset α)`?

Reply:
Yes. Though we don't have `subsingleton` and `nontrivial` for finset: what would you like me to use? (I think we should have them for finset but that wasn't a piece of work I was willing to commit to getting through.)

Reply:
Ah, if we don't have them then forget about it. It was just in case these lemmas were trivial generalizations.

Reply:
(I have submitted a commit with one possible fix - let me know what you think.)

Reply:
Essentially if you like I can leave the finset spelling in as-is, though it really is equivalent to these if we don't fix the coercions. I could also change it to `s.off_diag.nonempty` and `s.off_diag = \empty` if you like.